api/hasura/actions/upload_profile_avatar.ts

```diff
@@ -0,0 +1,96 @@
+import type { VercelRequest, VercelResponse } from '@vercel/node';
+import { v4 as uuidv4 } from 'uuid';
+
+import { gql } from '../../../api-lib/Gql';
+import { deleteImage, uploadImage } from '../../../api-lib/s3';
+import {
+  composeHasuraActionRequestBodyWithSession,
+  HasuraUserSessionVariables,
+  updateProfileAvatarInput,
+} from '../../../src/lib/zod';
+import { parseBase64Image, resizeAvatar } from '../images';
+
+export default async function handler(req: VercelRequest, res: VercelResponse) {
+  // parse the input
+  const {
+    input: { object: input },
+    session_variables: sessionVariables,
+  } = composeHasuraActionRequestBodyWithSession(
+    updateProfileAvatarInput,
+    HasuraUserSessionVariables
+  ).parse(req.body);
+
+  try {
+    const imageBytes = parseBase64Image(res, input.image_data_base64);
+    if (!imageBytes) {
+      return res;
+    }
+
+    // Figure out if there was a previous avatar, because we'll need to delete it
+    const { profiles_by_pk } = await gql.q('query')({
+      profiles_by_pk: [
+        {
+          id: sessionVariables.hasuraProfileId,
+        },
+        {
+          avatar: true,
+        },
+      ],
+    });
+
+    // see if there was a previous avatar so we can delete it after new one is saved
+    let previousAvatar: string | undefined = undefined;
+    if (profiles_by_pk) {
+      previousAvatar = profiles_by_pk.avatar;
+    }
```

Comment:
this is a nit, but can we do this after the graphql mutation completes?

Reply:
sure good call

Reply:
Actually no we can't, or I don't understand - we are doing the delete after the mutation succeeds, but this lookup has to happen beforehand or there is no way to grab the previousAvatar (it will be gone). Maybe you meant something else?